File: LeetCode/212.Word_Search_II/212.Word_Search_II.py
#!/usr/bin/env python3


class TrieNode:

    def __init__(self):
        self.children = {}
        self.word = None


class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        # Trie
        # current issue, how to handle so many path
        # 1. build trie by words
        # 2. dfs, search trie, if not, return

        m, n = len(board), len(board[0])
        root = TrieNode()

        def insert(word):
            cur = root
            for ch in word:
                if ch not in cur.children:
                    cur.children[ch] = TrieNode()
                cur = cur.children[ch]
            cur.word = word

        for w in words:
            insert(w)

        res = set()

        def dfs(i, j, node, cnt):
            ch = board[i][j]
            if ch not in node.children:
                return

            nxt = node.children[ch]
            if nxt.word:
                res.add(nxt.word)
                node.word = None

            if cnt > 10:
                return

            board[i][j] = "#"
            for di, dj in [(-1, 0), (0, -1), (1, 0), (0, 1)]:
                ni, nj = i + di, j + dj
                if 0 <= ni < m and 0 <= nj < n and board[ni][nj] != "#":
                    dfs(ni, nj, nxt, cnt + 1)
            board[i][j] = ch

        for i in range(m):
            for j in range(n):
                dfs(i, j, root, 0)

        return list(res)

File: LeetCode/212.Word_Search_II/212.Word_Search_II.java
class Solution {

    private int[][] dirs = { { 1, 0 }, { 0, 1 }, { -1, 0 }, { 0, -1 } };
    private Set<String> res = new HashSet<String>();

    public List<String> findWords(char[][] board, String[] words) {
        TrieNode root = new TrieNode();
        for (String word : words) {
            root.insert(word);
        }

        int m = board.length;
        int n = (board.length > 0) ? board[0].length : 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                dfs(i, j, m, n, board, root);
            }
        }

        return new ArrayList<String>(res);
    }

    private void dfs(int i, int j, int m, int n, char[][] board, TrieNode cur) {
        if (i < 0 || i >= m || j < 0 || j >= n) {
            return;
        }

        char ch = board[i][j];
        int index = ch - 'a';
        if (ch != '#' && cur.children[index] != null) {
            TrieNode node = cur.children[index];
            if (node.word != "") {
                res.add(node.word);
            }
            board[i][j] = '#';
            for (int[] dir : dirs) {
                dfs(i + dir[0], j + dir[1], m, n, board, node);
            }
            board[i][j] = ch;
        }
        return;
    }

}

class TrieNode {
    public String word;
    public TrieNode[] children;

    public TrieNode() {
        word = "";
        children = new TrieNode[26];
    }

    public void insert(String word) {
        TrieNode cur = this;
        for (int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);
            int index = ch - 'a';
            if (cur.children[index] == null) {
                cur.children[index] = new TrieNode();
            }
            cur = cur.children[index];
        }
        cur.word = word;
    }
}